### PR TITLE
Support external event loops via non-blocking poll() alt. to run()

### PR DESCRIPTION
**Pull request by jamesbeilby**
For use in applications that already have a primary event loop, so they cannot call blocking run(). They may instead call poll() within this loop to delegate to uWebSocket's secondary event loop (whether Asio, libuv or epoll). It is assumed that this will be called very frequently, e.g. "busy spin".

### CODE REVIEW

**Comment by alexhultman** [comment]
What kind of loop do you have in mind?

File: src/Node.cpp [hunk 1]
```diff
@@ -64,6 +64,10 @@ void Node::run() {
     loop->run();
 }
 
+void Node::poll() {
```
Comment (alexhultman) [inline src/Node.cpp]: makes sense

File: src/Libuv.h [hunk 2]
```diff
@@ -24,6 +24,10 @@ struct Loop : uv_loop_t {
     void run() {
         uv_run(this, UV_RUN_DEFAULT);
     }
+
+    void poll() {
```
Comment (alexhultman) [inline src/Libuv.h]: makes sense

File: src/Asio.h [hunk 3]
```diff
@@ -22,6 +22,10 @@ struct Loop : boost::asio::io_service {
     void run() {
         boost::asio::io_service::run();
     }
+
+    void poll() {
```
Comment (alexhultman) [inline src/Asio.h]: makes sense

File: src/Epoll.cpp [hunk 4]
```diff
@@ -9,55 +9,64 @@ std::recursive_mutex cbMutex;
...
+
+void Loop::run() {
+    while (numPolls) {
+    	doEpoll(delay);
+    }
+}
+
+void Loop::poll() {
+	if (numPolls) {
+		doEpoll(0);
```
Comment (alexhultman) [inline src/Epoll.cpp]: tabs here

File: src/Epoll.cpp [hunk 5]
```diff
@@ -9,55 +9,64 @@ std::recursive_mutex cbMutex;
...
     }
+
+    if (postCb) {
+        postCb(postCbData);
+    }
+}
+
+void Loop::run() {
+    while (numPolls) {
+    	doEpoll(delay);
```
Comment (alexhultman) [inline src/Epoll.cpp]: spaces and tabs here

File: src/Epoll.cpp [hunk 6]
```diff
@@ -9,55 +9,64 @@ std::recursive_mutex cbMutex;
 void (*callbacks[16])(Poll *, int, int);
 int cbHead = 0;
 
-void Loop::run() {
-    timepoint = std::chrono::system_clock::now();
```
Comment (alexhultman) [inline src/Epoll.cpp]: Why did you remove this? It breaks timers in some cases

**Comment by jamesbeilby** [comment]
Thanks Alex.

For context, I have a low latency application following Actor pattern (with message passing between threads), and this Actor framework uses a custom dispatch/event loop as it needs to support non-standard I/O (like user-space network sockets) - so ruling out epoll for example. It is useful to be able to handle HTTP/WS requests on one of these Actor threads, as the same thread can then send/receive messages to the rest of the application, without any synchronization.

I have fixed the tabs.

I was going to ask about timepoint -- I thought it possible that it was set conservatively, noting that it is updated prior to invoking callbacks (line 26). All the timers I could find are created either from within these callbacks or possibly before run() is called. Perhaps this is to handle case where numPolls == 0 when run is called, or I am just missing the point?

Cheers

**Comment by alexhultman** [comment]
> Perhaps this is to handle case where numPolls == 0 when run is called, or I am just missing the point?

Yep. It comes from strange libuv behavior:

http://docs.libuv.org/en/v1.x/loop.html#c.uv_update_time

Calling Hub::run() with no polls should update the loop's timepoint like is done in libuv to keep same behavior.

**Comment by jamesbeilby** [comment]
Makes sense thanks. Have reverted this.

**Comment by chronoxor** [comment]
Is it possible provide Hub/Node initialization with external Loop? For example I have my own asio::io_service and I want to use it with uWebSockets Hub.

**Comment by jamesbeilby** [comment]
chrono, it would be possible with this change to have your own io_service and uWebSockets' io_service running on the same thread, but this would be inefficient. I would suggest maybe a separate change that handles "defaultLoop = true" in Loop::Loop in Asio.h, or that extends defaultLoop to allow a custom loop to be constructed and passed.

**Comment by alexhultman** [comment]
Yes the idea is to support seamless integration with libuv & asio projects. You could possibly get the loop from µWS using Hub::getLoop() but otherwise you need to PR that change so that you can pass in any io_service (probably just make that a void ptr or something).

**Comment by chronoxor** [comment]
I'll try to create such functionality with external loop initialization after I finish with Asio Standalone loop, which works without boost dependency.

Injecting into the static createLoop() is a bad idia, because it is difficult to control workflow especially in multithread env.

**Comment by chronoxor** [comment]
Suddenly I found that uWebSockets lib does not support ASIO on Windows platform because the lack of boost::asio::posix::stream_descriptor.

Then I tried to replace it with boost::asio::ip::tcp::socket. Some modifications were made in Asio.h (see attachment). And I managed to compile and run echo.cpp sample. But when I connect to it with a client WebSocket from my Chrome I get Access Violation error in debugger.

[Asio.h.zip](https://github.com/uNetworking/uWebSockets/files/1700345/Asio.h.zip)
uwebsockets
